mac/hid.c

```diff
@@ -552,6 +552,36 @@ static struct hid_device_info *create_device_info_with_usage(IOHIDDeviceRef dev,
 		cur_dev->interface_number = -1;
 	}
 
+    if (cur_dev->interface_number == -1 || cur_dev->interface_number == 0) {
+        int old_interface_number = cur_dev->interface_number;
+        //try to Fallback to older interface number find rules
+        io_string_t temp_dev_path_str;
+        char* temp_dev_path = NULL;
+        /* Fill in the path (IOService plane) */
+        if (iokit_dev != MACH_PORT_NULL) {
+            res = IORegistryEntryGetPath(iokit_dev, kIOServicePlane, temp_dev_path_str);
+            if (res == KERN_SUCCESS)
+                temp_dev_path = strdup(temp_dev_path_str);
+            else
+                temp_dev_path = strdup("");
+        }
+
+        if (temp_dev_path) {
+            const char* match_prefix_string = "Interface@";
+            char* interface_component = strstr(temp_dev_path, match_prefix_string);
+            if (interface_component) {
+                char* decimal_str = interface_component + strlen(match_prefix_string);
```

Comment:
A string literal instead of a `const char *` would avoid potentially compiling in a runtime call to `strlen()`, as `sizeof("...") - 1` is equal to `strlen("...")` but is guaranteed to be evaluated compile-time. Doing `strlen(s)` on a `const char *s = "..."` _might_ get optimized away with some compiler flags, but it definitely won't always be.

Of course a string literal would have to be typed out twice, once for `strstr()` and then a second time for `sizeof()`. If you can stomach it you can define a temporary macro or two.

Because at this point I think I've written enough comment spam I'll just post an alternative to the whole code block, with a couple of additional minor tweaks in addition to the ones I've mentioned. Haven't compiled, no guarantees that it actually works. It just might.

```
	if (iokit_dev != MACH_PORT_NULL &&
	    (cur_dev->interface_number == -1 || cur_dev->interface_number == 0)) {
		io_string_t temp_dev_path_str;
		temp_dev_path_str[0] = '\0';

		/* Fill in the path (IOService plane) */
		res = IORegistryEntryGetPath(iokit_dev, kIOServicePlane, temp_dev_path_str);
		if (res == KERN_SUCCESS) {
			#define match_prefix_str "Interface@"
			#define match_prefix_len (sizeof(match_prefix_str) - 1)

			char *interface_component = strstr(temp_dev_path_str, match_prefix_str);
			if (interface_component) {
				char *decimal_str = interface_component + match_prefix_len;
				if (decimal_str[0]) {
					char *endptr = decimal_str;
					int interface_number = strtol(decimal_str, &endptr, 10);
					if (endptr != decimal_str) {
						/* Parsing succeeded, update the interface number. */
						cur_dev->interface_number = interface_number;
					}
				}
			}

			#undef match_prefix_len
			#undef match_prefix_str
		}
	}

```

Reply:
thanks for suggestion